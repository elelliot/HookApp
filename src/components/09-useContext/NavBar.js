import React from 'react'
import { Link, NavLink } from 'react-router-dom';

export const NavBar = () => {
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            <Link to="#" className="navbar-brand" >useContext</Link>
            
            <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                <div className="navbar-nav">
                    {/*Link es lo mismo que NavLink solo que este puede establecer una clase si el url coincide con el path*/}
                    <NavLink exact activeClassName="active" to="/" className="nav-link" >Home</NavLink>
                    <NavLink exact activeClassName="active" to="/about" className="nav-link" >About</NavLink>
                    <NavLink exact activeClassName="active" to="/login" className="nav-link" >Login</NavLink>
                </div>
            </div>
        </nav>
        // <nav>
        //     <ul>
        //         <li>
        //             {/* Los Links funcionan más rápido que un tag <a href>, necesitamos tener react router instalado en el proyecto y no hará refresh cuando entremos a un link*/}
        //             <Link to="/"> Home </Link>
        //         </li>
        //         <li>
        //             <Link to="/about"> About </Link>
        //         </li>
        //         <li>
        //             <Link to="/login"> Login </Link>
        //         </li>
                
        //     </ul>
        // </nav>
    )
}
