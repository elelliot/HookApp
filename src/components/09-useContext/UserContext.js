import { createContext } from 'react';

export const UserContext = createContext(null); //Creamos este componente para poder enviar valores desde el MainApp