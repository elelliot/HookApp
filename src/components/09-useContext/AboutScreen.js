import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const AboutScreen = () => {

    const {user, setUser} = useContext(UserContext);

    const handleClick = () => {
        setUser({}); //purgamos la informacion
    }

    return (
        <div>
            <h1>AboutScreen</h1>
            <hr />

            <pre>
                { JSON.stringify( user, null, 3)/*Visualizamos la info*/ }
            </pre>

            <button 
                className="btn btn-warning"
                onClick={ handleClick }
            >
                Logout
            </button>
        </div>
    )
}
