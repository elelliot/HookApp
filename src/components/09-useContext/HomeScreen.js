import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    const { user } = useContext(UserContext);//Recibimos el valor del UserContext que creamos, ser√≠a el value del MainApp.js el cual es un objeto, y lo desestructuramos
    // const just = useContext(UserContext);
    // console.log( just );
    console.log( user );

    return (
        <div>
            <h1>HomeScreen</h1>
            <hr />

            <pre>
                { JSON.stringify( user, null, 3) }
            </pre>
        </div>
    )
}
