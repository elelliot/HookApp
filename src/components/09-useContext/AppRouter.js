import React from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Redirect,
  } from "react-router-dom";

  import { NavBar } from './NavBar';
  import { HomeScreen } from './HomeScreen';
  import { AboutScreen } from './AboutScreen';
import { LoginScreen } from './LoginScreen';

export const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar />

                <div className="container">

                    {/* Esto es con react-router-dom v5, esto no funciona para la 6 */}
                    <Switch>
                        <Route exact path="/" component ={ HomeScreen }/>{/*Si una ruta es as√≠ con puro "/", te lleva al que cumpla ya que no es exacta la expresion regular, poner el exact para corregir eso */}

                        <Route exact path="/about" component ={ AboutScreen }/>{/*Disparamos el AboutScreen si el url termina con /about */ }
                        <Route exact path="/login" component ={ LoginScreen }/>
                        
                        {/*El Switch funciona como el condicional que conocemos, solo que los cases son el path del Router,
                        y cuando tenemos un caso que no coincide con ninguno, aplicamos un default (tal cual como un switch),
                        podemos hacerlo de 2 formas:
                        1- Usamos un route que no tiene path, y lo mandamos al componente que queremos
                        2- El Redirect redirecciona al path que eliges (default) en caso de que el url no coincida con ningun path */}
                        {/* <Route component = { HomeScreen } /> */}
                        <Redirect to ="/" />

                    </Switch>

                </div>


            </div>
        </Router>
    )
}
