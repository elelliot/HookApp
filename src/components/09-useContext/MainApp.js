import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'

export const MainApp = () => {
    
    const [user, setUser] = useState({});
    
    return (
        <UserContext.Provider value = {{ 
            user,
            setUser,
            }}>
            {/*Los Higher Order components son aquellos que tienen componentes hijos,
            todos los hijos del UserContext puede recibir el valor de user o cualquier otro valor/función
            (no se para que se pone .Provider, pero así es) */ }

            <AppRouter />

        </UserContext.Provider>
    )
}
