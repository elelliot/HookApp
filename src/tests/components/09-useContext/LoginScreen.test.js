import { mount } from "enzyme";
import { LoginScreen } from "../../../components/09-useContext/LoginScreen";
import { UserContext } from "../../../components/09-useContext/UserContext";


describe('Pruebas en <LoginScreen />', () => {

    

    //Declaramos el mock de la funci칩n
    const setUser = jest.fn();


    //Montamos el component
    const wrapper = mount( 
        <UserContext.Provider value={{
            setUser      
        }}>
            <LoginScreen />
        </UserContext.Provider>
     );



    test('Debe mostrarse bien', () => {
        //Snapshot
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe ejecutar el setUser con el argumento esperado', () => {
        
        //Buscamos el elemento y simulamos el click
        // wrapper.find('button').simulate('click'); esta forma est치 bien pero...
        
        //es m치s directo llamar al onClick, y como es funci칩n debemos llamarlo con los parentesis
        wrapper.find('button').prop('onClick')();

        //Confirmar si se llama la funcion...
        // expect( setUser ).toHaveBeenCalledTimes(1);

        //...con los args
        expect( setUser ).toHaveBeenCalledWith({
            id:123, name:'ElElliot'
        });

        
    });
    
    
    
})
