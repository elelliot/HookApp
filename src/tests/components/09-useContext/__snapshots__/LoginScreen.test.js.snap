// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Debe mostrarse bien 1`] = `
<div>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
